dnf5/main.cpp

```diff
@@ -825,6 +841,47 @@ static void dump_variables(Context & context) {
     }
 }
 
+static void print_new_leaves(Context & context) {
+    libdnf5::rpm::PackageQuery pkg_query(context.base);
+    pkg_query.filter_installed();
+    libdnf5::rpm::PackageQuery pre_trans_leaves_query(pkg_query);
+    pre_trans_leaves_query.filter_leaves();
+
+    // Calculate the new system state (list of installed packages) after a successful transaction.
+    // Current state plus inbound minus outbound packages.
+    for (const auto & trans_pkg : context.get_transaction()->get_transaction_packages()) {
+        if (libdnf5::transaction::transaction_item_action_is_inbound(trans_pkg.get_action())) {
+            pkg_query.add(trans_pkg.get_package());
+        } else if (libdnf5::transaction::transaction_item_action_is_outbound(trans_pkg.get_action())) {
+            pkg_query.remove(trans_pkg.get_package());
+        }
+    }
+
+    pkg_query.filter_leaves();
+    pkg_query.difference(pre_trans_leaves_query);
+
+    std::set<std::string> pre_trans_leaves_na;
+    std::set<std::string> new_leaves_na;
+    for (const auto package : pre_trans_leaves_query) {
+        pre_trans_leaves_na.insert(package.get_na());
+    }
+    for (const auto package : pkg_query) {
```

Comment:
Maybe rather loop using `const auto &`, also in the for cycle above?

Reply:
`PackageQuery` uses a bitmap internally. The package is created during the dereference of the iterator and is returned by value. So using `&` reference after `auto` does not save anything. It will be a reference to the object that will be created anyway. I wanted to emphasize that the Package objects are being created and intentionally did not put `&`.

But ok, I'll add `&`. In this case, the compiler should generate the same code after optimization. So it will be a cosmetic change.

Reply:
Thanks for the explanation. So it makes better sense to omit the `&` there, but as of now, we already have a lot of `const auto &` usages iterating through the package queries in our codebase, so at least we are keeping it consistent :slightly_smiling_face: